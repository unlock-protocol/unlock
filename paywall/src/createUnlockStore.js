import { createStore, applyMiddleware, combineReducers, compose } from 'redux'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { createMemoryHistory } from 'history'

import configure from './config'

// Reducers
import keysReducer, {
  initialState as defaultKeys,
} from './reducers/keysReducer'
import locksReducer, {
  initialState as defaultLocks,
} from './reducers/locksReducer'
import networkReducer, {
  initialState as defaultNetwork,
} from './reducers/networkReducer'
import providerReducer, {
  initialState as defaultProvider,
} from './reducers/providerReducer'
import transactionsReducer, {
  initialState as defaultTransactions,
} from './reducers/transactionsReducer'
import currencyReducer, {
  initialState as defaultCurrency,
} from './reducers/currencyReducer'
import errorsReducer, {
  initialState as defaultError,
} from './reducers/errorsReducer'
import accountReducer, {
  initialState as defaultAccount,
} from './reducers/accountReducer'
import modalReducer, {
  initialState as defaultModals,
} from './reducers/modalsReducer'
import walletStatusReducer, {
  initialState as defaultWalletStatus,
} from './reducers/walletStatusReducer'

const config = configure()

export const createUnlockStore = (
  defaultState = {},
  history = createMemoryHistory(),
  middlewares = []
) => {
  const reducers = {
    router: connectRouter(history),
    account: accountReducer,
    keys: keysReducer,
    locks: locksReducer,
    modals: modalReducer,
    network: networkReducer,
    provider: providerReducer,
    transactions: transactionsReducer,
    currency: currencyReducer,
    errors: errorsReducer,
    walletStatus: walletStatusReducer,
  }

  // Cleanup the defaultState to remove all null values so that we do not overwrite existing
  // values with null
  Object.keys(defaultState).forEach(
    k => defaultState[k] == null && delete defaultState[k]
  )

  // We build the initial state by taking first each reducer's default values
  // Then some overides and finally whatever state we have stored locally.
  const initialState = Object.assign(
    {
      account: defaultAccount,
      keys: defaultKeys,
      locks: defaultLocks,
      modals: defaultModals,
      network: defaultNetwork,
      provider: defaultProvider,
      transactions: defaultTransactions,
      currency: defaultCurrency,
      errors: defaultError,
      walletStatus: defaultWalletStatus,
    },
    {
      provider: Object.keys(config.providers)[0],
    },
    defaultState
  )

  middlewares.push(routerMiddleware(history))

  const composeEnhancers =
    (global.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
        trace: true,
        traceLimit: 25,
      })) ||
    compose

  return createStore(
    combineReducers(reducers),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  )
}

export default createUnlockStore
