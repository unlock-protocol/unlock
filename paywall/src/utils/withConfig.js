import React from 'react'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

export const ConfigContext = React.createContext()

/**
 * This creates an HOC from a component and injects the configuration.
 * It also triggers errors if constraints are not respected.
 * @param {*} Component
 */
export default function withConfig(Component) {
  function componentWithConfig(props) {
    return (
      <ConfigContext.Consumer>
        {config => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  componentWithConfig.getInitialProps = async context => {
    return {
      ...(Component.getInitialProps
        ? await Component.getInitialProps(context)
        : {}),
    }
  }

  return componentWithConfig
}
