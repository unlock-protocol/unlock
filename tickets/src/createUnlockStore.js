import { createStore, applyMiddleware, combineReducers, compose } from 'redux'

import configure from './config'

// Reducers
import locksReducer, {
  initialState as defaultLocks,
} from './reducers/locksReducer'
import loadingReducer, {
  initialState as defaultLoading,
} from './reducers/loadingReducer'
import networkReducer, {
  initialState as defaultNetwork,
} from './reducers/networkReducer'
import providerReducer, {
  initialState as defaultProvider,
} from './reducers/providerReducer'
import errorsReducer, {
  initialState as defaultError,
} from './reducers/errorsReducer'
import accountReducer, {
  initialState as defaultAccount,
} from './reducers/accountReducer'
import walletStatusReducer, {
  initialState as defaultWalletStatus,
} from './reducers/walletStatusReducer'
import eventReducer, {
  initialState as defaultEvent,
} from './reducers/eventReducer'

const config = configure()

export const createUnlockStore = (defaultState = {}, middlewares = []) => {
  const reducers = {
    account: accountReducer,
    locks: locksReducer,
    loading: loadingReducer,
    network: networkReducer,
    provider: providerReducer,
    errors: errorsReducer,
    walletStatus: walletStatusReducer,
    event: eventReducer,
  }

  // Cleanup the defaultState to remove all null values so that we do not overwrite existing
  // values with null
  Object.keys(defaultState).forEach(
    k => defaultState[k] == null && delete defaultState[k]
  )

  // We build the initial state by taking first each reducer's default values
  // Then some overides and finally whatever state we have stored locally.
  const initialState = {
    account: defaultAccount,
    locks: defaultLocks,
    loading: defaultLoading,
    network: defaultNetwork,
    provider: defaultProvider,
    errors: defaultError,
    walletStatus: defaultWalletStatus,
    event: defaultEvent,
    provider: Object.keys(config.providers)[0],
    ...defaultState,
  }

  const composeEnhancers =
    (global.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
        trace: true,
        traceLimit: 25,
      })) ||
    compose

  return createStore(
    combineReducers(reducers),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  )
}

export default createUnlockStore
