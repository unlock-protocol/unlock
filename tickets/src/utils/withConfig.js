import React from 'react'
import { connect } from 'react-redux'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

export const ConfigContext = React.createContext()

/**
 * This creates an HOC from a component and injects the configuration.
 * It also triggers errors if constraints are not respected.
 * @param {*} Component
 */
export default function withConfig(Component) {
  function componentWithConfig(props) {
    return (
      <ConfigContext.Consumer>
        {config => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  function mapStateToProps(state) {
    return {
      network: state.network,
      account: state.account,
    }
  }

  componentWithConfig.getInitialProps = async context => {
    return {
      ...(Component.getInitialProps
        ? await Component.getInitialProps(context)
        : {}),
    }
  }

  return connect(mapStateToProps)(componentWithConfig)
}
