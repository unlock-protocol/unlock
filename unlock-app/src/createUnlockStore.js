import { loadState } from './services/localStorageService'
import { createStore, applyMiddleware, combineReducers } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'

// Reducers
import networkReducer from './reducers/networkReducer'
import providerReducer from './reducers/providerReducer'
import transactionReducer  from './reducers/transactionReducer'
import locksReducer  from './reducers/locksReducer'
import keysReducer  from './reducers/keysReducer'

// Middlewares
import lockMiddleware from './middlewares/lockMiddleware'

export default function createUnlockStore(config, browserHistory) {
  const reducers = {
    provider: providerReducer,
    network: networkReducer,
    transactions: transactionReducer,
    locks: locksReducer,
    keys: keysReducer,
  }

  // TODO: DRY this because each reducer has its own initial state.
  const initialState = Object.assign({
    network: {
      name: 0,
    },
    transactions: {
      latest: null,
      all: {},
      lastUpdated: 0,
    },
    locks: {},
    keys: {},
  }, loadState())

  const middlewares = [
    lockMiddleware,
  ]

  return createStore(
    combineReducers(reducers),
    initialState,
    composeWithDevTools(applyMiddleware(...middlewares))
  )
}
