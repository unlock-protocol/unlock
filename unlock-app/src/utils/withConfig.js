import React from 'react'
import configure from '../config'
import { WrongNetwork, MissingProvider } from '../components/creator/FatalError'
import { ETHEREUM_NETWORKS_NAMES } from '../constants'
import { connect } from 'react-redux'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

const isServer = typeof window === 'undefined'
const config = configure(global)
const ConfigContext = React.createContext(config)

// This function takes a component...
export function withConfig(Component) {
  // ...and returns another component...
  function componentWithConfig(props) {
    // ... and renders the wrapped component with the context config!
    // Notice that we pass through any additional props as well

    const { store: reduxStore, router } = props

    if (!isServer) {
      // Ensuring that we have at least a provider
      if (Object.keys(config.providers).length === 0) {
        return (<MissingProvider />)
      }

      // Ensuring that the provider is using the right network!
      if (router.route !== '/provider' && config.isRequiredNetwork && !config.isRequiredNetwork(reduxStore.getState().network.name)) {
        return (<WrongNetwork currentNetwork={ETHEREUM_NETWORKS_NAMES[reduxStore.getState().network.name][0]} requiredNetwork={config.requiredNetwork} />)
      }
    }

    return (
      <ConfigContext.Consumer>
        {config => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  function mapStateToProps (state) {
    return {
      account: state.network.account, // TODO change account to base level
      network: state.network,
    }
  }

  componentWithConfig.getInitialProps = async context => {
    return {
      ...(Component.getInitialProps ? await Component.getInitialProps(context) : {}),
    }
  }

  return connect(mapStateToProps)(componentWithConfig)
}
