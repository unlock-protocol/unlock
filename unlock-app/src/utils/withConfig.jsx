import React, { useContext } from 'react'

/**
 * Function which creates higher order component with the config
 * Taken from https://reactjs.org/docs/context.html#consuming-context-with-a-hoc
 */

export const ConfigContext = React.createContext()

/**
 * This creates an HOC from a component and injects the configuration.
 * It also triggers errors if constraints are not respected.
 * @param {*} Component
 */
export default function withConfig(Component) {
  function componentWithConfig(props) {
    return (
      <ConfigContext.Consumer>
        {(config) => <Component {...props} config={config} />}
      </ConfigContext.Consumer>
    )
  }

  return componentWithConfig
}

export function useConfig() {
  return useContext(ConfigContext)
}
