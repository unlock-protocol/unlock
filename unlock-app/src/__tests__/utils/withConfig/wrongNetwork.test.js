import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import withConfig from '../../../utils/withConfig'
import createUnlockStore from '../../../createUnlockStore'

const store = createUnlockStore({
  network: {
    name: 4,
  },
})

const Component = () => <div>An unlock component</div>

const ComponentWithConfig = withConfig(Component)

jest.mock('../../../config', () =>
  jest.fn().mockImplementation(() => {
    return {
      providers: {
        HTTP: {},
      },
      isRequiredNetwork: () => false,
      requiredNetwork: 'dev',
    }
  })
)

describe('withConfig High Order Component', () => {
  describe('when the current network is not in the list of required networks', () => {
    it('should show a message indicating that the network needs to be changed', () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <ComponentWithConfig router={{ route: '/' }} />
          </Provider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
