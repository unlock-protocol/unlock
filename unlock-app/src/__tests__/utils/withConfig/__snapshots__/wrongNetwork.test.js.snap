// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withConfig High Order Component when the current network is not in the list of required networks should show a message indicating that the network needs to be changed 1`] = `
<div
  className="hm3mhg-0 gzhZVm"
>
  <div
    className="hm3mhg-1 hXUOmH"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <div
        className="cenpj1-0 laMZpD"
      >
        <svg
          viewBox="0 0 56 56"
        >
          <path
            d="M33.467 30.455c0 .365-.02.733-.07 1.095-.434 3.2-3.945 5.285-7.054 4.192-2.274-.799-3.7-2.769-3.7-5.134-.002-1.62-.002-3.241-.002-4.862h10.828c0 1.57 0 3.139-.002 4.709m6.719-14.07V16H33.41v6.61H22.58l.001-6.228v-.374h-6.753c-.006.08-.014.139-.014.197 0 2.135-.005 4.27-.008 6.405H14v3.14h1.806c.001 1.7.006 3.4.02 5.1.03 3.36 1.456 6.143 4.099 8.243 3.811 3.029 8.114 3.673 12.708 2.016 4.35-1.569 7.612-5.66 7.563-10.643a525.27 525.27 0 0 1-.023-4.716H42v-3.14h-1.825c.004-2.075.01-4.15.01-6.225"
          />
        </svg>
      </div>
    </a>
  </div>
  <div
    className="hm3mhg-3 cFpjmx"
  >
    <header
      className="s1glv5oz-0 lkbUqR"
    >
      <h1
        className="s1glv5oz-1 hxdRe"
      >
        Creator Dashboard
      </h1>
      <a
        className="s850ddk-0 lbEhWm"
        href="/about"
        onClick={[Function]}
      >
        <svg>
          <path
            d="M13.2 18h-2.4v-7.2h2.4V18zm0-9.6h-2.4V6h2.4v2.4z"
            fillRule="evenodd"
          />
        </svg>
        <small
          className="s850ddk-1 hAgZgI"
        >
          About
        </small>
      </a>
      <a
        className="s850ddk-0 lbEhWm"
        href="/jobs"
        onClick={[Function]}
      >
        <svg>
          <path
            d="M18.125 6.944c.482 0 .875.374.875.834V10H5V7.778c0-.46.393-.834.875-.834h3.5v-.833c0-.46.684-1.111 1.167-1.111h2.916c.483 0 1.167.652 1.167 1.111v.833h3.5zM16.083 11H19v5.1a.89.89 0 0 1-.875.9H5.875A.889.889 0 0 1 5 16.1V11h11.083z"
            fillRule="evenodd"
          />
        </svg>
        <small
          className="s850ddk-1 hAgZgI"
        >
          Join us
        </small>
      </a>
      <a
        className="s850ddk-0 lbEhWm"
        href="https://github.com/unlock-protocol/unlock"
        onClick={[Function]}
      >
        <svg>
          <path
            d="M9.01 23.977c.01-.177.016-.374.016-.591 0-.292-.01-1.066-.016-2.093-3.338.744-4.043-1.65-4.043-1.65-.545-1.42-1.332-1.798-1.332-1.798-1.09-.764.083-.749.083-.749 1.204.087 1.837 1.268 1.837 1.268 1.071 1.88 2.809 1.338 3.493 1.022.109-.795.42-1.337.762-1.645-2.665-.31-5.466-1.365-5.466-6.08 0-1.343.467-2.442 1.235-3.302-.123-.311-.535-1.562.117-3.256 0 0 1.008-.33 3.3 1.261a11.241 11.241 0 0 1 3.005-.414c1.019.005 2.046.141 3.004.414 2.29-1.592 3.296-1.261 3.296-1.261.655 1.694.244 2.945.12 3.256.77.86 1.233 1.959 1.233 3.302 0 4.726-2.805 5.767-5.478 6.071.43.38.814 1.13.814 2.279 0 1.645-.015 2.971-.015 3.375v.591H9.01z"
            fillRule="evenodd"
          />
        </svg>
        <small
          className="s850ddk-1 hAgZgI"
        >
          Source Code
        </small>
      </a>
    </header>
    <section
      className="s1g6piw8-0 dSRnSS"
    >
      <img
        className="s1g6piw8-1 fregfm"
        src="/static/images/illustrations/network.svg"
      />
      <div
        className="s1g6piw8-2 csrrBV"
      >
        <h1>
          Network mismatch
        </h1>
        <p>
          Youâ€™re currently on the 
          Rinkeby
           network but you need to be on the 
          dev
           network. Please switch to 
          dev
          .
        </p>
      </div>
    </section>
  </div>
  <div
    className="hm3mhg-2 ftYCdu"
  />
</div>
`;
